Add new LSP extension 'metals/pickInput' to implement the "Create Scala file" command

alekseiAlefirov opened the pull request:
After new-scala-file command has been added to metals https://github.com/scalameta/metals/pull/1339 , client-side service was discussed https://github.com/scalameta/metals-vscode/pull/183 , and it was decided to move and implemet quick-pick communication to server.

alekseiAlefirov commented on metals/src/main/scala/scala/meta/internal/metals/NewFilesProvider.scala:
```diff
@@ -7,51 +7,120 @@ import MetalsEnrichments._
+      .orElse(focusedDocument.map(_.parent))
+
+    val newlyCreatedFile =
+      askForKind
```
I've made some kind of DSL here, to cover `Future[Option[_]]` handling (making this method more or readable). `Future` is for about remote requests, and `Option` reflect that the request can be cancelled (thus, nothing should be returned eventually).
Not sure, if it's good here.

alekseiAlefirov commented on tests/unit/src/main/scala/tests/TestingClient.scala:
```diff
@@ -71,6 +71,7 @@ final class TestingClient(workspace: AbsolutePath, buffers: Buffers)
     _: ShowMessageRequestParams =>
       None
   }
+  var inputBoxHandler: MetalsInputBoxParams => Option[MetalsInputBoxResult] = _
```
while picking kind of file can be expressed via `showMessageRequest`, asking for name of the new file cannot, so I had to add this handler for testing purposes.

tgodzik replied:
Can you add a default implementation returning some random name? You can have `var nameToPick = "Foo"` and replace that if needed. You also have None as the default if it's an issue.

Otherwise, we will get a null pointer in case of some particular changes.

alekseiAlefirov commented on metals/src/main/scala/scala/meta/internal/metals/MetalsServerConfig.scala:
```diff
@@ -106,6 +111,8 @@ object MetalsServerConfig {
           openFilesOnRenames = true,
           executeClientCommand = ExecuteClientCommandConfig.on,
           globSyntax = GlobSyntaxConfig.vscode,
+          isInputBoxEnabled = true,
```
I had to change it as well, because entering name for the new file cannot be expressed via `showMessageRequest`. This resulted into a bit changed UI when formatting and no scalafmt version is set.
Also, I would expect for such configs to be set by client, but okay, I mimicked it with `isPickInputEnabled`.

tgodzik replied:
I would even remove it from here, since we want to drop the properties.

ckipp01 commented on docs/editors/new-editor.md:
```diff
@@ -134,6 +134,14 @@ Possible values:
   Metals tries to fallback to `window/showMessageRequest` when possible.
 - `on`: the `metals/inputBox` request is fully supported.
 
+### `-Dmetals.pick-input`
```
Related to some of the work I started in here: https://github.com/scalameta/metals/pull/1414, is it possible to not introduce this as a property, but rather as a `clientExperimentalCapabilities`? I feel like this would fit well there.

alekseiAlefirov replied:
Agreed, as I said in my previous comment, I would expect it to be a client config, rather then a server one. I've just repeated after `input-box`. If it's to be changed -- then this is to be changed as well :)

tgodzik replied:
Let's use the recently merged way defining capabilities.

tgodzik commented on tests/unit/src/test/scala/tests/NewFilesLspSuite.scala:
```diff
@@ -2,17 +2,36 @@ package tests
     Files.createDirectories(
       workspace.resolve("a/src/main/scala/").toNIO
     )
+    client.showMessageRequestHandler = { params =>
```
Maybe a utility closure:
```
def createMessageRequestHander(name : String) : MessageRequestParams => Option[MessageRequestResult] =
{ params =>
      if (NewScalaFile.isSelectTheKindOfFile(params)) {
        params.getActions().asScala.find(_.getTitle() == name)
      } else {
        None
      }
    }
```

then you can just use:
`client.showMessageRequestHandler = createMessageRequestHander("worksheet") `

tgodzik replied:
Or even better create a check method:
```
def check(picked : String, expectedOutput : String, name : String = "Foo", shouldCheckForName : Boolean)
```
then you can just have:
```
check(
  "object",
 """| object Foo {
     |}
     |""".stripMargin
```

createMessageRequestHande might not be needed then.

alekseiAlefirov replied:
not sure, if I got you correctly, but I've done something there - please, check.

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/NewFilesProvider.scala:
```diff
@@ -66,31 +135,80 @@ class NewFilesProvider(
+
+  implicit class DialogContinuation[A](state: Future[Option[A]]) {
+
+    def continueWith[B](
```
Looks good, this is almost a MonadTransformer haha.

Maybe `flatMapOption` and `mapOption` ? This will be a bit more natural.

And these methods might be actually useful to add to `MetalsEnrichments`, I remember situations when we were combining Futures and Options.

alekseiAlefirov replied:
Yeah, I also thought about `OptionT` :D . Done

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/NewFilesProvider.scala:
```diff
@@ -66,31 +135,80 @@ class NewFilesProvider(
     val result = Future {
-      Files.createFile(path.toNIO).toUri()
+      AbsolutePath(
+        Files.createFile(path.toNIO)
```
You can use `touch` from `XtensionAbsolutePathBuffers`

alekseiAlefirov replied:
`touch` does not return the path, and although I already have the path, I like to reuse output of the function.

tgodzik replied:
Can we change touch to return the path?

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/NewFilesProvider.scala:
```diff
@@ -7,51 +7,120 @@ import MetalsEnrichments._
+  )(
+      implicit ec: ExecutionContext
+  ): Future[AbsolutePath] = {
+    val path = directory.getOrElse(workspace).resolve(name + ".scala")
```
What happens if the name is `a/b/c/A` - I think we should create the directories there. Most likely `createFile` should also create directories? Package will then be added automatically which is neat.

alekseiAlefirov replied:
it fails. Not really sure, it looks neat to ask for the name for the new class `foo/Bar` to create `class Bar`at `foo/Bar.scala`

tgodzik replied:
Could we implement it here? I think we just need to do `getParent.mkdirs`

alekseiAlefirov commented on docs/editors/new-editor.md:
```diff
@@ -529,6 +537,73 @@ export interface MetalsInputBoxResult {
+  /**
+   * An optional flag to include the description when filtering the picks.
+   */
+  matchOnDescription?: boolean;
```
good question to discuss.
I've added them to match, yes, to be able to use the most of VSCode API.
But if we don't want them - we can remove them, they'll be initialized to default values on VSCode side.

tgodzik replied:
It should be fine, we can leave it as is. Unless they might cause issues for implementing this in other clients?

alekseiAlefirov replied:
I think, it won't be no problem, thanks to JS agility. Even now it got implicitly converted from Metals interface to VSCode one.

ckipp01 commented on metals/src/main/scala/scala/meta/internal/metals/MetalsServerConfig.scala:
```diff
@@ -106,6 +111,8 @@ object MetalsServerConfig {
           executeClientCommand = ExecuteClientCommandConfig.on,
           globSyntax = GlobSyntaxConfig.vscode,
+          isInputBoxEnabled = true,
+          isPickInputEnabled = true,
```
As a follow-up to my previous comment, and similiar to what @tgodzik mentioned, if you rebase on master now, you'll see the pattern I followed in #1414. Moving forward for _most cases_, we shouldn't need to add new server properties, but rather list them under `ClientExperimentalCapabilities`.

alekseiAlefirov replied:
I've added client properies, but server ones are still there, cause `https://github.com/scalameta/metals-vscode/pull/207` is not being pushed.

ckipp01 replied:
But that's because those properties already exist and are relied upon via clients that are already using them. This is a brand new one that we are introducing, so I'm not sure I understand why we also need to support this with client properties since nothing is relying on it? We can just only support it via `ClientExperimentalCapabilities`, and introduce it in vscode via experimentl.

alekseiAlefirov replied:
`isPickInputEnabled` removed now, and capability added to the client in https://github.com/scalameta/metals-vscode/pull/183. Though `isInputBoxEnabled = true` still in this place, I am not sure, if it should not be here now.

ckipp01 replied:
Yea,  for now that's fine since that property already existed anyways. Once we release, and the the vs code pr gets merged, we can then remove the `isInputBoxEnabled` since then it will be set via experimental from vs code.

alekseiAlefirov commented on metals/src/main/scala/scala/meta/internal/metals/ConfiguredLanguageClient.scala:
```diff
@@ -117,4 +129,13 @@ final class ConfiguredLanguageClient(
     }
   }
 
+  private implicit def metalsPickInputParams2ShowMessageRequestInputParams(
```
done, changed to
```
private def toShowMessageRequestParams(
      params: MetalsPickInputParams
  ): ShowMessageRequestParams
```

alekseiAlefirov commented on metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageClient.scala:
```diff
@@ -112,3 +123,33 @@ case class MetalsInputBoxResult(
     @Nullable cancelled: java.lang.Boolean = null
 )
+
+case class MetalsPickInputParams(
```
I just followed `MetalsInputBoxParams`, defined in this file before.

alekseiAlefirov commented on metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:
```diff
@@ -1266,27 +1273,18 @@ class MetalsLanguageServer(
+              Some(directory.getAsString()).map(new URI(_))
+            )
+          case Seq(_: JsonNull) =>
+            newFilesProvider.createNewFileDialog(None)
```
done

alekseiAlefirov commented on metals/src/main/scala/scala/meta/internal/metals/NewFilesProvider.scala:
```diff
@@ -66,31 +135,80 @@ class NewFilesProvider(
+      new ExecuteCommandParams(
+        ClientCommands.GotoLocation.id,
+        List(
+          new Location(path.toString(), new org.eclipse.lsp4j.Range()): Object
```
`Range` fixed.
Good catch with `toURI`, thanks!
`:Object` can not be ommitted, since it's Java List eventually, and it's not covariant. It's the same with already existing cases of calling `ClientCommands.GotoLocation`

alekseiAlefirov commented on metals/src/main/scala/scala/meta/internal/metals/Messages.scala:
```diff
@@ -382,4 +382,15 @@ class Messages(icons: Icons) {
+    def selectTheKindOfFileMessage = "Select the kind of file to create"
+    def enterNameMessage(kind: String): String = s"Enter name for the new $kind"
+
+    def isSelectTheKindOfFile(params: ShowMessageRequestParams): Boolean =
```
It's the same with other `is...` methods in this file - they are used only in tests.

tgodzik replied:
That's what I meant, I think they can just be move to the test class.

alekseiAlefirov commented on metals/src/main/scala/scala/meta/internal/metals/Messages.scala:
```diff
@@ -382,4 +382,15 @@ class Messages(icons: Icons) {
+
+  object NewScalaFile {
+    def selectTheKindOfFileMessage = "Select the kind of file to create"
+    def enterNameMessage(kind: String): String = s"Enter name for the new $kind"
```
fixed

alekseiAlefirov commented on metals/src/main/scala/scala/meta/internal/metals/NewFilesProvider.scala:
```diff
@@ -7,51 +7,120 @@ import MetalsEnrichments._
+      name: String,
+      kind: String
+  )(
+      implicit ec: ExecutionContext
```
done

olafurpg commented on docs/editors/new-editor.md:
````diff
@@ -541,6 +541,73 @@ export interface MetalsInputBoxResult {
 }
 ```
 
+### `metals/pickInput`
````
Would it make sense to name this `metals/quickPick`? In general I try to follow the same naming convention as in the VS Code APIs

alekseiAlefirov replied:
Yeah, I was not sure about the name, so if you think, this one's better, okay then. Renamed.

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/MetalsEnrichments.scala:
```diff
@@ -602,4 +602,25 @@ object MetalsEnrichments
+      )
+  }
+
+  implicit class OptionFutureLift[A](state: Future[A]) {
```
There should already be `XtensionScalaFuture`

alekseiAlefirov replied:
done

tgodzik commented on tests/unit/src/test/scala/tests/NewFilesLspSuite.scala:
```diff
@@ -155,26 +132,64 @@ class NewFilesLspSuite extends BaseLspSuite("new-files") {
+    } yield ()
+  }
 
+  private def check(
```
I like the direction, but I thought of something similar to RenameLspSuite and other, where we would even invoke `test` from here.

alekseiAlefirov replied:
done

alekseiAlefirov commented:
@tgodzik 
>> isInputBoxEnabled = true,

> I would even remove it from here, since we want to drop the properties.

What should I do with the client part https://github.com/scalameta/metals-vscode/pull/183 ?
- I could add `inputBox` capability, but it's already in @ckipp01 's PR https://github.com/scalameta/metals-vscode/pull/207
- I could rebase on that PR, so it is adequate to this PR without the `isInputBoxEnabled` line
- I could leave it, just wait for that PR to be merged.

alekseiAlefirov commented:
@tgodzik PR reworked,
> That's what I meant, I think they can just be move to the test class.

Moved.
> I would even remove it from here, since we want to drop the properties.

removed, but see previous comment.

> Could we implement it here? I think we just need to do getParent.mkdirs

done.

> Can we change touch to return the path?

Nah, I just returned path there (and changed `Files.create` to `touch`).

ckipp01 commented on metals/src/main/scala/scala/meta/internal/metals/Messages.scala:
```diff
@@ -382,4 +382,10 @@ class Messages(icons: Icons) {
+  object NewScalaFile {
+    def selectTheKindOfFileMessage = "Select the kind of file to create"
+    def enterNameMessage(kind: String): String =
+      s"Enter the name for the new $kind"
```
Super nitpicky, but I just tested this, and is it possible to just add a `:` in here?
```suggestion
      s"Enter the name for the new $kind:"
```

alekseiAlefirov replied:
it seems not to be going look good here:
![image](https://user-images.githubusercontent.com/10850363/73563467-adfe6880-445d-11ea-89f6-a9a6398034da.gif)

tgodzik replied:
Let's leave it as is, I don't think it adds that much visually and it's not really being used in UI anywhere.

ckipp01 replied:
Hmmm, yea I see what you're saying. That's tricky, because in `coc-metals` there isn't a nice break, so it seems like there should really be a `:`. No worries, I'll just add it in on the vim side.

<img width="438" alt="Screenshot 2020-02-24 at 14 01 42" src="https://user-images.githubusercontent.com/13974112/75154558-3b0ba900-570e-11ea-81fa-d6ef9b2fbfd5.png">

ckipp01 replied:
Also, this feature works great 🎉 !!

alekseiAlefirov replied:
thanks :relieved:
